File: C++/ABC/ABC008/D_99.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int w, h, n;
    cin >> w >> h >> n;
    vector<int> x(n), y(n);
    for (int i = 0; i < n; i++) {
        cin >> x[i] >> y[i];
    }
    int res = 0;
    for (int i = 0; i < n; i++) {
        res |= (1 << i);
    }
    auto dfs = [&](auto self, int m, int a, int b, int c, int d) -> int {
        if (!m) return 0;
        vector<int> v;
        for (int i = 0; i < n; i++) {
            if ((m >> i) & 1) v.push_back(i);
        }
        int res = 0;
        for (auto i: v) {
            int p = 0, q = 0, r = 0, s = 0;
            for (auto j: v) {
                if (x[j] < x[i] && y[j] < y[i]) p |= (1 << j);
                if (x[j] < x[i] && y[j] > y[i]) q |= (1 << j);
                if (x[j] > x[i] && y[j] < y[i]) r |= (1 << j);
                if (x[j] > x[i] && y[j] > y[i]) s |= (1 << j);
            }
            int cr = c - a + d - b + 1;
            cr += self(self, p, a, b, x[i] - 1, y[i] - 1);
            cr += self(self, q, a, y[i] + 1, x[i] - 1, d);
            cr += self(self, r, x[i] + 1, b, c, y[i] - 1);
            cr += self(self, s, x[i] + 1, y[i] + 1, c, d);
            res = max(res, cr);
        }
        return res;
    };
    cout << dfs(dfs, res, 1, 1, w, h) << '\n';
}

File: Python/ABC/ABC008/D_99.py
w, h = map(int, input().split())
n = int(input())
t = [tuple(map(int, input().split())) for _ in range(n)]
ans = 0
for i in range(n):
    ans |= (1 << i)
def dfs(m, a, b, c, d):
    if m == 0:
        return 0
    v = []
    for i in range(n):
        if (m >> i) & 1:
            v.append(i)
    res = 0
    for i in v:
        p, q, r, s = 0, 0, 0, 0
        xi, yi = t[i]
        for j in v:
            xj, yj = t[j]
            if xj < xi and yj < yi:
                p |= (1 << j)
            if xj < xi and yj > yi:
                q |= (1 << j)
            if xj > xi and yj < yi:
                r |= (1 << j)
            if xj > xi and yj > yi:
                s |= (1 << j)
        cr = c - a + d - b + 1 + dfs(p, a, b, xi - 1, yi - 1) + dfs(q, a, yi + 1, xi - 1, d) + dfs(r, xi + 1, b, c, yi - 1) + dfs(s, xi + 1, yi + 1, c, d)
        res = max(res, cr)
    return res
print(dfs(ans, 1, 1, w, h))